core/src/fork_selector.rs

```diff
@@ -0,0 +1,288 @@
+use crate::replay_stage_ds::ForkProgress;
+use solana_runtime::bank::Bank;
+use solana_sdk::clock::Slot;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+#[derive(Clone)]
+pub struct Fork {
+    pub heaviest_slot: Slot,
+    pub fork_contents: Vec<Slot>,
+}
+
+pub trait ForkSelector {
+    /// extracts fork structure from set of frozen banks,
+    /// ascendants and descendants
+    fn extract_forks(
+        &self,
+        frozen_banks: &HashMap<Slot, Arc<Bank>>,
+        ancestors: &HashMap<u64, HashSet<u64>>,
+        descendants: &HashMap<u64, HashSet<u64>>,
+        progress: &HashMap<u64, ForkProgress>,
+    ) -> Vec<Fork>;
+
+    /// finds best fork out of vector of fork
+    fn find_best_fork(
+        &self,
+        forks: Vec<Fork>,
+        progress: &HashMap<u64, ForkProgress>,
+    ) -> Option<Fork>;
+}
+
+#[derive(Default)]
+pub struct OptimalForkSelector {}
+
+impl ForkSelector for OptimalForkSelector {
+    fn extract_forks(
+        &self,
+        frozen_banks: &HashMap<Slot, Arc<Bank>>,
+        ancestors: &HashMap<u64, HashSet<u64>>,
+        descendants: &HashMap<u64, HashSet<u64>>,
+        progress: &HashMap<u64, ForkProgress>,
+    ) -> Vec<Fork> {
+        let mut forks: Vec<Fork> = vec![];
+
+        for (_i, slot) in ancestors.keys().enumerate() {
```

Comment:
where is this logic coming from?

Reply:
its hard to verify that this code does the same thing as the current code.  I was expecting the implementation to match what is currently in replay stage
````
    fn select_fork(                                                
        my_pubkey: &Pubkey,     
        ancestors: &HashMap<u64, HashSet<u64>>,       
        bank_forks: &Arc<RwLock<BankForks>>,                               
        tower: &Tower,                                                   
        progress: &mut HashMap<u64, ForkProgress>,                  
    ) -> VoteAndPoHBank 
```

Reply:
also, you don't need the `enumerate` if you are ignoring the `_i` parameter

Reply:
ah I see, the select_fork function is renamed to compute_fork_progress